### PR TITLE
MF-170 - Add manager API unit tests

### LINKED ISSUE #170: Raise the test coverage
We have tests only for the mainflux/manager and mainflux/manager/api. I think we should write more tests in order to be able to change things without worrying that we'll break something.

I will start writing some tests. and I want to start with the http and writer services. I will use the same tools/packages for testing which are used in the [manager_test.go](https://github.com/mainflux/mainflux/blob/master/manager/manager_test.go).

What do you think about this? Are there any parts that it does not make sense to write tests for?

Best Regards,
Jovan

### PR DESCRIPTION
**Pull request by anovakovic01**
This pull request resolves #170.

**Summary:**
- Add unit test for every manager API endpoint
- Update swagger documentation
- Fix connection request validation
- Fix EOF error handling
- Add invalid content type error handling
- Update README with codecov badge

Signed-off-by: Aleksandar Novakovic <anovakovic01@gmail.com>

### CODE REVIEW

**Review by drasko — DISMISSED** [review]
LGTM

**Comment by chombium** [comment]
@anovakovic01 this looks cool, thanks for your work.
One question: I don't see the pact file in the commit. What are your thoughts on storing the pact files?
We need them available somewhere in order to verify the expected behavior of the "service provider".

I guess they are stored in the pact directory which you've added to the .gitignore.
I'll checkout your branch,  play a bit and give my comments.

Best Regards,
Jovan

**Comment by anovakovic01** [comment]
@chombium Thanks for feedback. 

I didn't commit pact files because they are generated by running tests in `consumer_test`, so I think that you have everything you need to run pact tests. Also, if I am not mistaken, pact files are usually shared over pact broker which we won't need in our case (at least for now). We are not using pact in standard way because we don't have consumer service. Instead I wrote test use cases in `consumer_test` and generated pact files from them. If you think there are any reasons we should commit pact files, let me know.

Best Regards,
Aleksandar

File: manager/api/transport_test.go [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+	svc := newService()
+	mux := api.MakeHandler(svc)
+	ts := httptest.NewServer(mux)
+	defer ts.Close()
+	client := ts.Client()
+
+	test(svc, ts.URL, client)
+}
+
+func sendRequest(client *http.Client, method string, url string, contentType string, token string, body io.Reader) (*http.Response, error) {
```
Comment (mijicd) [inline manager/api/transport_test.go]: Please rename to `makeRequest`, and wrap all of the parameters in a structure called `testRequest`.

File: manager/api/transport_test.go [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+
+		cases := []struct {
+			desc   string
+			req    string
+			status int
+		}{
+			{"register new user", string(validBody), http.StatusCreated},
+			{"try to register existing user", string(validBody), http.StatusConflict},
+			{"try to register user with invalid email address", string(invalidBody), http.StatusBadRequest},
+			{"try to register with invalid request format", "{", http.StatusBadRequest},
```
Comment (mijicd) [inline manager/api/transport_test.go]: Empty request would be viable test case as well.

File: manager/api/transport_test.go [hunk 3]
```diff
@@ -0,0 +1,535 @@
...
+		validBody, _ := json.Marshal(validUser)
+		invalidBody, _ := json.Marshal(invalidUser)
+
+		cases := []struct {
+			desc   string
+			req    string
+			status int
+		}{
+			{"register new user", string(validBody), http.StatusCreated},
+			{"try to register existing user", string(validBody), http.StatusConflict},
```
Comment (mijicd) [inline manager/api/transport_test.go]: Please drop "try to". This applies for all specs.

File: manager/api/transport_test.go [hunk 4]
```diff
@@ -0,0 +1,535 @@
...
+	users := mocks.NewUserRepository()
+	clients := mocks.NewClientRepository()
+	channels := mocks.NewChannelRepository()
+	hasher := mocks.NewHasher()
+	idp := mocks.NewIdentityProvider()
+
+	return manager.New(users, clients, channels, hasher, idp)
+}
+
+func runService(test func(manager.Service, string, *http.Client)) {
```
Comment (mijicd) [inline manager/api/transport_test.go]: Instead of passing a test callback, rename this method to `newServer` and return its instance in the same way it is done at `newService`. You can obtain all of the data in the test itself.

Just to be clear, this is not a bad way to remove duplication, but I'd prefer that we are as explicit as possible in our test cases.

File: manager/api/transport_test.go [hunk 5]
```diff
@@ -0,0 +1,535 @@
...
+		nonexistentCredentials, _ := json.Marshal(nonexistentUser)
+		svc.Register(validUser)
+
+		cases := []struct {
+			desc   string
+			req    string
+			status int
+			res    string
+		}{
+			{"login with valid credentials", string(validCredentials), http.StatusCreated, string(token)},
```
Comment (mijicd) [inline manager/api/transport_test.go]: These `string(...)` are hard to read. Don't "torture" the table, do the cast during test data preparation. This applies to each of the further occurrences.

File: manager/api/transport_test.go [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+	"github.com/mainflux/mainflux/manager/api"
+	"github.com/mainflux/mainflux/manager/mocks"
+	"github.com/stretchr/testify/assert"
+)
+
+const contentType = "application/json; charset=utf-8"
+
+var (
+	validUser       = manager.User{"user@example.com", "password"}
+	invalidPassUser = manager.User{"user@example.com", "invalid_password"}
```
Comment (mijicd) [inline manager/api/transport_test.go]: Since this is used only once, you can drop it from here and create an instance locally, based on the `validUser`.

File: manager/api/transport_test.go [hunk 7]
```diff
@@ -0,0 +1,535 @@
...
+	invalidUser     = manager.User{"userexample.com", "password"}
```
Comment (mijicd) [inline manager/api/transport_test.go]: This "fellow" exists for the sole purpose of providing a false e-mail value. Define a constant that contains "bad" value instead (this can be reused for other purposes as well, see other tests as an example).

The only exception to the aforementioned statement is user registration. For that purpose you can create local instance based on the `validUser` but using the constant I mentioned as an e-mail.

File: manager/api/transport_test.go [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+	nonexistentUser = manager.User{"non-existentuser@example.com", "pass"}
+	validClient     = manager.Client{Type: "app", Name: "test_app", Payload: "test_payload"}
+	invalidClient   = manager.Client{Type: "foo", Name: "invalid_client", Payload: "some_payload"}
```
Comment (mijicd) [inline manager/api/transport_test.go]: Same remark as `invalidUser` and `invalidPassUser`.

File: manager/api/transport_test.go [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+	"github.com/mainflux/mainflux/manager"
+	"github.com/mainflux/mainflux/manager/api"
+	"github.com/mainflux/mainflux/manager/mocks"
+	"github.com/stretchr/testify/assert"
+)
+
+const contentType = "application/json; charset=utf-8"
+
+var (
+	validUser       = manager.User{"user@example.com", "password"}
```
Comment (mijicd) [inline manager/api/transport_test.go]: Please drop "valid" prefix from all valid entities. This will improve readability together with other changes.

File: manager/api/transport_test.go [hunk 10]
```diff
@@ -0,0 +1,535 @@
...
+func sendRequest(client *http.Client, method string, url string, contentType string, token string, body io.Reader) (*http.Response, error) {
+	req, err := http.NewRequest(method, url, body)
+	if err != nil {
+		return nil, err
+	}
+	if token != "" {
+		req.Header.Set("Authorization", token)
+	}
+	if contentType != "" {
+		req.Header.Set("Content-Type", contentType)
```
Comment (mijicd) [inline manager/api/transport_test.go]: Add test cases that cover the missing content type. Since this might be the edge case, we should expect `415` error.

File: manager/api/transport_test.go [hunk 11]
```diff
@@ -0,0 +1,535 @@
...
+			res, err := sendRequest(client, http.MethodPost, fmt.Sprintf("%s/clients", url), contentType, tc.auth, strings.NewReader(tc.req))
+			assert.Nil(t, err, fmt.Sprintf("%s: unexpected error %s", tc.desc, err))
+			assert.Equal(t, tc.status, res.StatusCode, fmt.Sprintf("%s: expected status code %d got %d", tc.desc, tc.status, res.StatusCode))
+		}
+	})
+}
+
+func TestUpdateClient(t *testing.T) {
+	runService(func(svc manager.Service, url string, client *http.Client) {
+		validUpdate, _ := json.Marshal(manager.Client{
```
Comment (mijicd) [inline manager/api/transport_test.go]: Why do you create new "valid" instance. Reuse!

File: manager/api/transport_test.go [hunk 12]
```diff
@@ -0,0 +1,535 @@
...
+			Type:    "device",
+			Name:    "test_device",
+			Payload: "device_payload",
+		})
+		invalidUpdate, _ := json.Marshal(manager.Client{
```
Comment (mijicd) [inline manager/api/transport_test.go]: Create invalid based on the existing valid. This will make the "error" more explicit.

File: manager/api/transport_test.go [hunk 13]
```diff
@@ -0,0 +1,535 @@
...
+			Type:    "foo",
+			Name:    "test_device",
+			Payload: "device_payload",
+		})
+		svc.Register(validUser)
+		clientID, _ := svc.AddClient(validUser.Email, validClient)
```
Comment (mijicd) [inline manager/api/transport_test.go]: Please rename to `id`, it is sufficient.

File: manager/api/transport_test.go [hunk 14]
```diff
@@ -0,0 +1,535 @@
...
+			assert.Nil(t, err, fmt.Sprintf("%s: unexpected error %s", tc.desc, err))
+			assert.Equal(t, tc.status, res.StatusCode, fmt.Sprintf("%s: expected status code %d got %d", tc.desc, tc.status, res.StatusCode))
+		}
+	})
+}
+
+func TestViewClient(t *testing.T) {
+	runService(func(svc manager.Service, url string, client *http.Client) {
+		svc.Register(validUser)
+		clientID, _ := svc.AddClient(validUser.Email, validClient)
```
Comment (mijicd) [inline manager/api/transport_test.go]: Please rename to `id`.

File: manager/api/transport_test.go [hunk 15]
```diff
@@ -0,0 +1,535 @@
...
+			Name:    "test_device",
+			Payload: "device_payload",
+		})
+		svc.Register(validUser)
+		clientID, _ := svc.AddClient(validUser.Email, validClient)
+
+		cases := []struct {
+			desc     string
+			req      string
+			clientID string
```
Comment (mijicd) [inline manager/api/transport_test.go]: Please rename to `ID`.

File: manager/api/transport_test.go [hunk 16]
```diff
@@ -0,0 +1,535 @@
...
+			auth     string
+			status   int
+		}{
+			{"update existing client", string(validUpdate), clientID, validUser.Email, http.StatusOK},
+			{"try to update non-existent client", string(validUpdate), "123e4567-e89b-12d3-a456-000000000042", validUser.Email, http.StatusNotFound},
```
Comment (mijicd) [inline manager/api/transport_test.go]: Wrong uuid can be made constant instead of repeating it.

File: manager/api/transport_test.go [hunk 17]
```diff
@@ -0,0 +1,535 @@
...
+			clients = append(clients, client)
+		}
+
+		cases := []struct {
+			desc   string
+			auth   string
+			status int
+			res    []manager.Client
+		}{
+			{"fetch list of clients", validUser.Email, http.StatusOK, clients},
```
Comment (mijicd) [inline manager/api/transport_test.go]: Maybe add another case where we retrieve clients for user without any clients (you should create that user as well).

File: manager/api/transport_test.go [hunk 18]
```diff
@@ -0,0 +1,535 @@
...
+			assert.Equal(t, tc.status, res.StatusCode, fmt.Sprintf("%s: expected status code %d got %d", tc.desc, tc.status, res.StatusCode))
+			assert.ElementsMatch(t, tc.res, data["clients"], fmt.Sprintf("%s: expected body %s got %s", tc.desc, tc.res, data["clients"]))
+		}
+	})
+}
+
+func TestRemoveClient(t *testing.T) {
+	runService(func(svc manager.Service, url string, client *http.Client) {
+		svc.Register(validUser)
+		clientID, _ := svc.AddClient(validUser.Email, validClient)
```
Comment (mijicd) [inline manager/api/transport_test.go]: Rename to `id`.

File: manager/api/transport_test.go [hunk 19]
```diff
@@ -0,0 +1,535 @@
...
+
+		cases := []struct {
+			desc   string
+			req    string
+			auth   string
+			status int
+		}{
+			{"create new channel", string(validData), validUser.Email, http.StatusCreated},
+			{"try to create new channel with invalid token", string(validData), invalidUser.Email, http.StatusForbidden},
+			{"try to create new channel with invalid data format", "{", validUser.Email, http.StatusBadRequest},
```
Comment (mijicd) [inline manager/api/transport_test.go]: Please add the empty JSON test.

File: manager/api/transport_test.go [hunk 20]
```diff
@@ -0,0 +1,535 @@
...
+		cases := []struct {
+			desc   string
+			req    string
+			auth   string
+			status int
+		}{
+			{"add valid client", string(validClient), validUser.Email, http.StatusCreated},
+			{"try to add client with invalid data", string(invalidClient), validUser.Email, http.StatusBadRequest},
+			{"try to add client with invalid auth token", string(validClient), "invalid_token", http.StatusForbidden},
+			{"try to add client with invalid request format", "}", validUser.Email, http.StatusBadRequest},
```
Comment (mijicd) [inline manager/api/transport_test.go]: Please add the empty JSON test.

File: manager/api/transport_test.go [hunk 21]
```diff
@@ -0,0 +1,535 @@
...
+			clientID string
+			auth     string
+			status   int
+		}{
+			{"update existing client", string(validUpdate), clientID, validUser.Email, http.StatusOK},
+			{"try to update non-existent client", string(validUpdate), "123e4567-e89b-12d3-a456-000000000042", validUser.Email, http.StatusNotFound},
+			{"try to update client with invalid id", string(validUpdate), "1", validUser.Email, http.StatusNotFound},
+			{"try to update client with invalid data", string(invalidUpdate), clientID, validUser.Email, http.StatusBadRequest},
+			{"try to update client with invalid user token", string(validUpdate), clientID, invalidUser.Email, http.StatusForbidden},
+			{"try to update client with invalid data format", "{", clientID, validUser.Email, http.StatusBadRequest},
```
Comment (mijicd) [inline manager/api/transport_test.go]: Please add the empty JSON test.

File: manager/api/transport_test.go [hunk 22]
```diff
@@ -0,0 +1,535 @@
...
+		cases := []struct {
+			desc     string
+			clientID string
+			auth     string
+			status   int
+		}{
+			{"delete existing client", clientID, validUser.Email, http.StatusNoContent},
+			{"try to delete non-existent client", "123e4567-e89b-12d3-a456-000000000042", validUser.Email, http.StatusNoContent},
+			{"try to delete client with invalid id", "1", validUser.Email, http.StatusNoContent},
+			{"try to delte client with invalid token", clientID, invalidUser.Email, http.StatusForbidden},
```
Comment (mijicd) [inline manager/api/transport_test.go]: Fix typo.

File: manager/api/transport_test.go [hunk 23]
```diff
@@ -0,0 +1,535 @@
...
+	})
+}
+
+func TestUpdateChannel(t *testing.T) {
+	runService(func(svc manager.Service, url string, client *http.Client) {
+		update, _ := json.Marshal(map[string]string{
+			"name": "updated_channel",
+		})
+		svc.Register(validUser)
+		chanID, _ := svc.CreateChannel(validUser.Email, validChannel)
```
Comment (mijicd) [inline manager/api/transport_test.go]: Remarks given for client-related tests regarding asking to rename to `id` apply here as well.

File: manager/api/transport_test.go [hunk 24]
```diff
@@ -0,0 +1,535 @@
...
+			req    string
+			chanID string
+			auth   string
+			status int
+		}{
+			{"update existing channel", string(update), chanID, validUser.Email, http.StatusOK},
+			{"try to update non-existing channel", string(update), "123e4567-e89b-12d3-a456-000000000042", validUser.Email, http.StatusNotFound},
+			{"try to update channel with invalid token", string(update), chanID, invalidUser.Email, http.StatusForbidden},
+			{"try to update channel with invalid id", string(update), "1", validUser.Email, http.StatusNotFound},
+			{"try to update channel with invalid data format", "}", chanID, validUser.Email, http.StatusBadRequest},
```
Comment (mijicd) [inline manager/api/transport_test.go]: Please add the empty JSON test.

File: manager/api/transport_test.go [hunk 25]
```diff
@@ -0,0 +1,535 @@
...
+		cases := []struct {
+			desc     string
+			chanID   string
+			clientID string
+			auth     string
+			status   int
+		}{
+			{"connect existing client to existing channel", chanID, clientID, validUser.Email, http.StatusOK},
+			{"try to connect existing client to non-existent channel", "123e4567-e89b-12d3-a456-000000000042", clientID, validUser.Email, http.StatusNotFound},
+			{"try to connect client with invalid id to channel", chanID, "1", validUser.Email, http.StatusNotFound},
```
Comment (mijicd) [inline manager/api/transport_test.go]: Add tests that will try to connect client from user A to the channel from user B and vice versa.

File: manager/api/transport_test.go [hunk 26]
```diff
@@ -0,0 +1,535 @@
...
+		svc.Connect(validUser.Email, chanID, clientID)
+
+		cases := []struct {
+			desc     string
+			chanID   string
+			clientID string
+			auth     string
+			status   int
+		}{
+			{"disconnect connected client from channel", chanID, clientID, validUser.Email, http.StatusNoContent},
```
Comment (mijicd) [inline manager/api/transport_test.go]: Add tests that try to disconnect channel from user A from client from user B and vice versa. This is an upgrade to the invalid token test :)

File: manager/api/transport_test.go [hunk 27]
```diff
@@ -0,0 +1,535 @@
...
+			assert.Equal(t, tc.clientID, clientID, fmt.Sprintf("%s: expected %s got %s", tc.desc, tc.clientID, clientID))
+		}
+	})
+}
+
+func TestCanAccess(t *testing.T) {
+	runService(func(svc manager.Service, url string, client *http.Client) {
+		svc.Register(validUser)
+		clientID, _ := svc.AddClient(validUser.Email, validClient)
+		unconnectedClientID, _ := svc.AddClient(validUser.Email, validClient)
```
Comment (mijicd) [inline manager/api/transport_test.go]: Please rename to `notConnectedClientID`.

File: manager/api/transport_test.go [hunk 28]
```diff
@@ -0,0 +1,535 @@
...
+		svc.Connect(validUser.Email, chanID, clientID)
+
+		cases := []struct {
+			desc      string
+			chanID    string
+			clientKey string
+			status    int
+			clientID  string
+		}{
+			{"validate token over existing channel id", chanID, clientID, http.StatusOK, clientID},
```
Comment (mijicd) [inline manager/api/transport_test.go]: Change descriptions to use the following "template":  `check access to [...] given [...]`.

File: manager/mocks/channels.go [hunk 29]
```diff
@@ -29,7 +28,7 @@ func (crm *channelRepositoryMock) Save(channel manager.Channel) (string, error)
 	defer crm.mu.Unlock()
 
 	crm.counter += 1
-	channel.ID = strconv.Itoa(crm.counter)
+	channel.ID = fmt.Sprintf("123e4567-e89b-12d3-a456-%012d", crm.counter)
```
Comment (mijicd) [inline manager/mocks/channels.go:33]: Theoretically, this will start giving invalid UUIDs after 9 entities. 

Instead, let's be more explicit about the way we want to use this implementation. Predefine an array of n valid uuids, and cycle through it using the counter modulo n as an index. Yes, there will be duplicates, but this should be a stub anyhow, and later on we can improve its logic if it is necessary.

File: manager/mocks/clients.go [hunk 30]
```diff
@@ -29,7 +28,7 @@ func (crm *clientRepositoryMock) Id() string {
 	defer crm.mu.Unlock()
 
 	crm.counter += 1
-	return strconv.Itoa(crm.counter)
+	return fmt.Sprintf("123e4567-e89b-12d3-a456-%012d", crm.counter)
```
Comment (mijicd) [inline manager/mocks/clients.go:31]: See the remark given for channels mock.

**Review by mijicd — CHANGES_REQUESTED** [review]

File: manager/api/transport_test.go (hunk 10 above)
Comment (anovakovic01) [reply manager/api/transport_test.go]: If content type is missing, handlers ignore that fact and work normally, so there is no need for this test case.

File: manager/mocks/channels.go (hunk 29 above)
Comment (mijicd) [reply manager/mocks/channels.go:33]: Please ignore this comment. I misunderstood your implementation.

File: manager/api/transport_test.go [hunk 31]
```diff
@@ -0,0 +1,718 @@
...
+
+	return manager.New(users, clients, channels, hasher, idp)
+}
+
+func newServer(svc manager.Service) *httptest.Server {
+	mux := api.MakeHandler(svc)
+	return httptest.NewServer(mux)
+}
+
+func makeRequest(client *http.Client, testReq testRequest) (*http.Response, error) {
```
Comment (mijicd) [inline manager/api/transport_test.go]: Even better approach is to bind this method, and the `http.Client` to the `testRequest`. Its signature would be `testRequest.make()`.

File: manager/api/transport_test.go [hunk 32]
```diff
@@ -0,0 +1,718 @@
...
+	}
+	return client.Do(req)
+}
+
+func TestRegister(t *testing.T) {
+	svc := newService()
+	ts := newServer(svc)
+	defer ts.Close()
+	client := ts.Client()
+	url := ts.URL
```
Comment (mijicd) [inline manager/api/transport_test.go]: Drop this variable.

File: manager/api/transport_test.go [hunk 33]
```diff
@@ -0,0 +1,718 @@
...
+
+	body, _ := json.Marshal(user)
```
Comment (mijicd) [inline manager/api/transport_test.go]: Please create helper `toJson(data interface{}) string`, and use it instead of creating "artificial" temporary names.

**Review by mijicd — CHANGES_REQUESTED** [review]
Please correct these minor changes along with the explicitly given remark:
- add `Content-Type` check in code and documentation
- try to integrate with codecov.io

**Review by mijicd — APPROVED** [review]

**Comment by chombium** [comment]
cool :) well done @anovakovic01 @mijicd
